### ISSUE TITLE
Apparent PP/UO dependency issue

### ISSUE THREAD

**Issue by jasondhales** [idaholab/moose#8485]
### Description of the enhancement or error report
In Bison, we found a few tests that were behaving oddly when we attempted to remove the legacy PP/UO option.  Cody mentioned there may be a dependency being missed.

We were able to find a way to get the tests to pass with the legacy option off, but we suspect that if things were executed in the proper order that we'd see diffs.  In fact, the tests seem to be computing the wrong values, and we expect diffs once the dependency is sorted out.

### Rationale for the enhancement or information for reproducing the error
To test, add `execute_on = 'initial timestep_end'` to the `he_prod` PP in bison/tests/ifba_he_production/ifba_examp_template.i and run the ifba tests.

### Identified impact
Proper sequence of PP/UO evaluation.

**Comment by jasondhales** [comment]
This test is run by setting cli_args.  See the tests file.

**Comment by permcody** [comment]
... Thanks!
Where is the UO though?

**Comment by jasondhales** [comment]
The UO comes in through an Action.  I know at least PlenumPressure creates one.

**Comment by permcody** [comment]
OK - Thanks

**Comment by permcody** [comment]
@gardnerru stopped by and said that BISON is working now without any legacy computation. I looked and I see that @acasagran has updated these HE production tests. However, there may still be some sort of missing dependency but I'm just not seeing it. Part of the issue here is that these files are extremely complex with roughly a dozen different PPs, and a few UserObjects. Some of these objects are depended on by AuxKernels which throws them into a different bucket altogether. If there's then a dependency on that PP or UO, the value can be lagged.

Specifically, this case cannot be automatically handled by MOOSE currently:
Aux -> PP (or UO) -> PP (or UO).

The reason is that the Aux dependency will cause the first PP to be executed earlier so that the values are current for the Aux. However we don't detect the transitive dependency so the second PP doesn't get properly moved with the previous PP. We have several open issues about intersystem dependency resolution but these are significant efforts that we just haven't had the time or focus to resolve. A year ago, I made significant changes to the MooseInterfaces to move towards detecting these kinds of dependencies: https://github.com/idaholab/moose/pull/6656. We still aren't there though. Even if the work in that previous PR gets finished we still have a lot more work to actually resolve them during a solve.

Yes we have plans to do this but it probably isn't going to happen without a work package to support it. It's just too big of an effort and will take a resource a very long time to complete. What I suggest is that in these cases, we have to try to work in the system. The example here can be resolved by simply having a dummy Aux depend on the second PP. In other cases, moving Execute on to `TIMESTEP_BEGIN` for some PPs while leaving `TIMESTEP_END` can help. I'm happy to advise while we wait for a better solution.

**Comment by YaqiWang** [comment]
One suggestion in this case is to explicitly couple another PP in the Aux although Aux is not directly in use of it. This is what we did in Rattlesnake. It can be an easy fix on the application side.

**Comment by YaqiWang** [comment]
@permcody Since this issue rises again, can you address my pull request at the very bottom of moose PR list? It can help BISON guys to locate the defect at least.

**Comment by permcody** [comment]
Yeah - That's what PR #6656 was working towards. It is theoretically detectable now. I just need to prioritize it. I'm swamped.

**Comment by GiudGiud** [comment]
Where are we on this?
I suspect this was addressed, the dependency resolution has changed a lot
@jasondhales
